src/test/java/com/fittracker/fittracker/service/WeightServiceTest.java

```diff
@@ -0,0 +1,92 @@
+
+import java.time.LocalDate;
+import java.util.Optional;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+public class WeightServiceTest {
+
+    @Mock
+    private WeightRepository weightRepository;
+
+    @InjectMocks
+    private WeightService weightService;
+
+    private final Weight weight = new Weight(LocalDate.of(2023, 10, 10), 100.1);
+
+    private final WeightRequest weightRequest = new WeightRequest(LocalDate.of(2023,10,10),100.1);
+
+    @Nested
+    class FindByDate {
+        @Test
+        void givenDateFound_shouldReturnWeightResponse() {
+            when(weightRepository.findByDate(any())).thenReturn(Optional.of(weight));
+            var expectedWeightResponse = WeightResponse.fromWeight(weight);
+            var result = weightService.findByDate(LocalDate.of(2023, 10, 10));
+
+            assertThat(result).usingRecursiveComparison().isEqualTo(expectedWeightResponse);
+            verify(weightRepository).findByDate(LocalDate.of(2023, 10, 10));
+        }
+
+        @Test
+        void givenNoDateFound_shouldThrowWeightNotFoundException() {
+            when(weightRepository.findByDate(any())).thenReturn(Optional.empty());
+
+            assertThatThrownBy(() -> weightService.findByDate(LocalDate.of(2023, 10, 10)))
+                    .isInstanceOf(WeightNotFoundException.class)
+                    .hasMessageContaining("Weight not found for date: 2023-10-10" );
+
+            verify(weightRepository).findByDate(LocalDate.of(2023, 10, 10));
+        }
+
+    }
+
+    @Nested
+    class Save {
+        @Test
+        void givenNoDateFound_shouldSaveAndReturnWeightResponse() {
+            when(weightRepository.existsByDate(any())).thenReturn(false);
+            when(weightRepository.save(any())).thenReturn(weightRequest.toWeight());
+
+            var expectedWeightResponse = WeightResponse.fromWeight(weightRequest.toWeight());
+            var result = weightService.save(weightRequest);
+
+            assertThat(result).usingRecursiveComparison().isEqualTo(expectedWeightResponse);
+            verify(weightRepository, times(1)).existsByDate(LocalDate.of(2023,10,10));
+            verify(weightRepository, times(1)).save(any());
+
+        }
+        @Test
+        void givenDateFound_shouldThrowExceptionAndStop() {
+            when(weightRepository.existsByDate(any())).thenReturn(true);
+
+            assertThatThrownBy(()-> weightService.save(weightRequest))
+                    .isInstanceOf(WeightAlreadyExistsException.class)
+                    .hasMessageContaining("Weight already exists for date: 2023-10-10");
+
+            verify(weightRepository,never()).save(any());
```

Comment:
we should verify that `existsByDate` was called, and then we can use `verifyNoMoreInteractions`